network/local.go

```diff
@@ -0,0 +1,459 @@
+}
+
+func newConnQueue() *connQueue {
+	return &connQueue{
+		Cond: sync.NewCond(&sync.Mutex{}),
+	}
+}
+
+func (c *connQueue) Push(p Packet) {
+	c.L.Lock()
+	defer c.L.Unlock()
+	if c.isClosed {
+		return
+	}
+	c.queue = append(c.queue, p)
+	c.Signal()
+}
+
+func (c *connQueue) Pop() (Packet, error) {
+	c.L.Lock()
+	defer c.L.Unlock()
+	for len(c.queue) == 0 {
+		if c.isClosed {
+			return EmptyApplicationPacket, ErrClosed
+		}
+		c.Wait()
+	}
+	if c.isClosed {
+		return EmptyApplicationPacket, ErrClosed
+	}
+	nm := c.queue[0]
+	c.queue = c.queue[1:]
+	return nm, nil
+}
+
+func (c *connQueue) Close() {
+	c.L.Lock()
+	defer c.L.Unlock()
+	c.isClosed = true
+	c.Signal()
+}
+
+// LocalListener is a Listener that uses LocalConn to communicate. It tries to
+// behave as much as possible as a real golang net.Listener but using LocalConn
+// as the underlying communication layer.
+type LocalListener struct {
+	// addr is the addr we're listening to + mut
+	addr Address
+	// are we listening or not
+	listening bool
+
+	sync.Mutex
+
+	// quit is used to stop the listening routine
+	quit chan bool
+
+	ctx *LocalContext
+}
+
+// NewLocalListener returns a fresh LocalListener which implements the Listener
+// interface.
+func NewLocalListener(addr Address) (*LocalListener, error) {
+	return NewLocalListenerWithContext(defaultLocalContext, addr)
+}
+
+// NewLocalListenerWithContext is similar to NewLocalListener but taking a
+// specific LocalContext to use to communicate.
+func NewLocalListenerWithContext(ctx *LocalContext, addr Address) (*LocalListener, error) {
+	l := &LocalListener{
+		quit: make(chan bool),
+		ctx:  ctx,
+	}
+	return l, l.bind(addr)
+}
+
+func (ll *LocalListener) bind(addr Address) error {
+	ll.Lock()
+	defer ll.Unlock()
+	if addr.ConnType() != Local {
+		return errors.New("Wrong address type for local listener")
+	}
+	if ll.ctx.IsListening(addr) {
+		return fmt.Errorf("%s is already listening: can't listen again", addr)
+	}
+	ll.addr = addr
+	return nil
+}
+
+// Listen implements the Listener interface
+func (ll *LocalListener) Listen(fn func(Conn)) error {
+	ll.Lock()
+	ll.quit = make(chan bool)
+	ll.ctx.Listening(ll.addr, fn)
+	ll.listening = true
+	ll.Unlock()
+
+	<-ll.quit
+	return nil
+}
+
+// Stop implements the Listener interface
+func (ll *LocalListener) Stop() error {
+	ll.Lock()
+	defer ll.Unlock()
+	ll.ctx.StopListening(ll.addr)
+	if ll.listening {
+		close(ll.quit)
+	}
+	ll.listening = false
+	return nil
+}
+
+// Address returns the listening address used.
+func (ll *LocalListener) Address() Address {
+	ll.Lock()
+	defer ll.Unlock()
+	return ll.addr
+}
+
+// Listening returns true if this Listener is actually listening for any
+// incoming connections.
+func (ll *LocalListener) Listening() bool {
+	ll.Lock()
+	defer ll.Unlock()
+	return ll.listening
+}
+
+// LocalHost is a Host implementation using LocalConn and LocalListener as
+// the underlying means of communication. It is a implementation of the Host
+// interface.
```

Comment:
I don't think so, Golang has this particularity of having kinda like the Ruby duck typing system but at compile time. So defining a struct don't tell you what interfaces it implements. Only by looking at which methods does this struct implements, you see that it has all the methods `Host` is defining **so** it is a `Host`.